README.md
# MkRefs

The **MkRefs** [plugin](http://www.mkdocs.org/user-guide/plugins/)
for [`MkDocs`](https://www.mkdocs.org/) 
generates reference Markdown pages from a knowledge graph, 
based on the [`kglab`](https://github.com/DerwenAI/kglab) project.

No graph database is required; however, let us know if you'd like to
use one in particular.

There are several planned use cases for the **MkRefs** plugin,
including:

  * *biblio* – semantic bibliography entries, generated from RDF
  * *glossary* – semantic glossary entries, generated from RDF
  * *apidocs* – semantic [*apidocs*](https://pypi.org/search/?q=apidocs) supporting the [Diátaxis](https://derwen.ai/docs/kgl/learn/#a-grammar-of-learning) grammar for documentation, generated as RDF from Python source code
  * *depend* – semantic dependency graph for Python libraries, generated as RDF from `setup.py`
  * *index* – semantic search index, generated as RDF from MkDocs content

Only the *apidocs*, *biblio*, and *glossary* components have been
added to **MkRefs** so far, although the other mentioned components
exist in separate projects and are being integrated.


<details>
  <summary>Contributing Code</summary>

We welcome people getting involved as contributors to this open source
project!

For detailed instructions please see:
[CONTRIBUTING.md](https://github.com/DerwenAI/mkrefs/blob/main/CONTRIBUTING.md)
</details>

<details>
  <summary>Semantic Versioning</summary>

Before <strong>MkRefs</strong> reaches release <code>v1.0.0</code> the 
types and classes may undergo substantial changes and the project is 
not guaranteed to have a consistent API.

Even so, we'll try to minimize breaking changes.
We'll also be sure to provide careful notes.

See:
[changelog.txt](https://github.com/DerwenAI/mkrefs/blob/main/changelog.txt)
</details>

<img
 alt="MkRefs, for semantic references"
 src="https://raw.githubusercontent.com/DerwenAI/mkrefs/main/docs/assets/logo.png"
 width="267"
/>


## Why does this matter?

A key takeaway is that many software engineering aspects of open
source projects involve graphs, therefore a knowledge graph can
provide an integral part of an open source repository.
Moreover, by using semantic representation (RDF) projects that
integrate with each other can share (i.e., federate) common resources,
for example to share definitions, analyze mutual dependencies, etc.


## Installation

To install the plugin using `pip`:

```
python3 -m pip install mkrefs
```

Then add the plugin into the `mkdocs.yml` file:
```yaml
plugins:
  - mkrefs
```
In addition, the following configuration parameter is expected:

  * `mkrefs_config` - YAML configuration file for **MkRefs**; e.g., `mkrefs.yml`

---

## API Docs

An `apidocs` parameter within the configuration file expects four
required sub-parameters:

 * `page` – name of the generated Markdown page, e.g., `ref.md`
 * `template` – a [Jinja2 template](https://jinja.palletsprojects.com/en/3.0.x/) to generate Markdown, e.g., `ref.jinja`
 * `package` – name of the package being documented
 * `git` – base URL for source modules in Git, e.g., `https://github.com/DerwenAI/mkrefs/blob/main`

There is an optional `includes` parameter, as a list of class
definitions to include.
If this is used, then all other classes get ignored.

See the source code in this repo for examples of how to format
Markdown within *docstrings*.
Specifically see the parameter documentation per method or function,
which differs slightly from pre-exisiting frameworks.

Note that the name of the generated Markdown page for the
apidocs must appear in the `nav` section of your `mkdocs.yml`
configuration file.
See the structure used in this repo for an example.

### Best Practices: RDF representation

You can use this library outside of MkDocs, i.e., calling it
programmatically, to generate an RDF graph to represent your package
API reference:

```
package_name = "mkrefs"
git_url = "https://github.com/DerwenAI/mkrefs/blob/main"
includes = [ "MkRefsPlugin", "PackageDoc" ]

pkg_doc = PackageDoc(package_name, git_url, includes)
pkg_doc.build()

kg = pkg_doc.get_rdf()
```

The `PackageDoc.get_rdf()` method returns an RDF graph as an instance
of an `kglab.KnowledgeGraph` object.
For more details, see <https://derwen.ai/docs/kgl/>


## Bibliography

A `biblio` parameter within the configuration file expects four
required sub-parameters:

 * `graph` – an RDF graph represented as a Turtle (TTL) file, e.g., `mkrefs.ttl`
 * `page` – name of the generated Markdown page, e.g., `biblio.md`
 * `template` – a [Jinja2 template](https://jinja.palletsprojects.com/en/3.0.x/) to generate Markdown, e.g., `biblio.jinja`
 * `queries` – [SPARQL queries](https://rdflib.readthedocs.io/en/stable/intro_to_sparql.html) used to extract bibliography data from the knowledge graph

See the [`mkrefs.ttl`](https://github.com/DerwenAI/mkrefs/blob/main/docs/mkrefs.ttl)
file for an example bibliography represented in RDF.
This comes from the documentation for the [`pytextrank`](https://derwen.ai/docs/ptr/biblio/)
open source project.

In the example RDF, the [*bibo*](http://bibliontology.com/) vocabulary represents
bibliographic entries, and the [*FOAF*](http://xmlns.com/foaf/spec/) vocabulary
represents authors.
This also uses two custom OWL relations from the [*derwen*](https://derwen.ai/ns/v1)
vocabulary:

  * `derw:citeKey` – citekey used to identify a bibliography entry within the documentation
  * `derw:openAccess` – open access URL for a bibliography entry (if any)

The `queries` parameter has three required SPARQL queries:

  * `entry` – to select the identifiers for all of the bibliograpy entries
  * `entry_author` – a mapping to identify author links for each bibliography entry
  * `entry_publisher` - the publisher link for each bibliography entry (if any)

Note that the name of the generated Markdown page for the
bibliography must appear in the `nav` section of your `mkdocs.yml`
configuration file.
See the structure used in this repo for an example.

You may use any valid RDF representation for a bibliography.
Just be sure to change the three SPARQL queries and the Jinja2
template accordingly.

While this example uses an adaptation of the
[MLA Citation Style](https://www.easybib.com/guides/citation-guides/mla-format/mla-citation/),
feel free to modify the Jinja2 template to generate whatever
bibliographic style you need.


### Best Practices: constructing bibliographies

As much as possible, bibliography entries should use the conventions at
<https://www.bibsonomy.org/>
for their [*citation keys*](https://bibdesk.sourceforge.io/manual/BibDeskHelp_2.html).

Journal abbreviations should use
[*ISO 4*](https://en.wikipedia.org/wiki/ISO_4) standards, 
for example from <https://academic-accelerator.com/Journal-Abbreviation/System>

Links to online versions of cited works should use
[DOI](https://www.doi.org/)
for [*persistent identifiers*](https://www.crossref.org/education/metadata/persistent-identifiers/).

When available, 
[*open access*](https://peerj.com/preprints/3119v1/)
URLs should be listed as well.


### What is going on here?

For example with the bibliography use case, when the plugin runs...

1. It parses its configuration file to identify the target Markdown page to generate and the Jinja2 template
2. The plugin also loads an RDF graph from the indicated TTL file
3. Three SPARQL queries are run to identify the unique entities to extract from the graph
4. The graph is serialized as [JSON-LD](https://derwen.ai/docs/kgl/ref/#kglab.KnowledgeGraph.save_jsonld)
5. The `author`, `publisher`, and bibliography `entry` entities are used to *denormalize* the graph into a JSON data object
6. The JSON is rendered using the Jinja2 template to generate the Markdown
7. The Markdown page is parsed and rendered by MkDocs as HTML, etc.


## Glossary

A `glossary` parameter within the configuration file expects four
required sub-parameters:

 * `graph` – an RDF graph represented as a Turtle (TTL) file, e.g., `mkrefs.ttl`
 * `page` – name of the generated Markdown page, e.g., `glossary.md`
 * `template` – a [Jinja2 template](https://jinja.palletsprojects.com/en/3.0.x/) to generate Markdown, e.g., `glossary.jinja`
 * `queries` – [SPARQL queries](https://rdflib.readthedocs.io/en/stable/intro_to_sparql.html) used to extract glossary data from the knowledge graph

See the [`mkrefs.ttl`](https://github.com/DerwenAI/mkrefs/blob/main/docs/mkrefs.ttl)
file for an example glossary represented in RDF.
This example RDF comes from documentation for the
[`pytextrank`](https://derwen.ai/docs/ptr/glossary/)
open source project.

In the example RDF, the [*cito*](http://purl.org/spar/cito/)
vocabulary represents citations to locally represented bibliographic
entries.
The [*skos*](http://www.w3.org/2004/02/skos/core#) vocabulary
provides support for [*taxonomy*](http://accidental-taxonomist.blogspot.com/)
features, e.g., semantic relations among glossary entries.
This example RDF also uses a definition from the
[*derwen*](https://derwen.ai/ns/v1) vocabulary:

  * `derw:Topic` – a `skos:Concept` used to represent glossary entries

The `queries` parameter has three required SPARQL queries:

  * `entry` – to select the identifiers for all of the bibliograpy entries
  * `entry_syn` – a mapping of synonyms (if any)
  * `entry_ref` – a mapping of external references (if any)
  * `entry_cite` – citations to the local bibliography citekeys (if any)
  * `entry_hyp` – a mapping of [*hypernyms*](https://en.wikipedia.org/wiki/Hyponymy_and_hypernymy) (if any)

Note that the name of the generated Markdown page for the glossary
must appear in the `nav` section of your `mkdocs.yml` configuration
file.
See the structure used in this repo for an example.

You may use any valid RDF representation for a glossary.
Just be sure to change the three SPARQL queries and the Jinja2
template accordingly.


## Usage

The standard way to generate documentation with MkDocs is:
```
mkdocs build
```

If you'd prefer to generate reference pages programmatically using
Python scripts, see the code for usage of the `MkRefsPlugin` class,
plus some utility functions:

  * `load_kg()`
  * `render_apidocs()`
  * `render_biblio()`
  * `render_glossary()`

There are also command line *entry points* provided, which can be
helpful during dev/test cycles on the semantic representation of your
content:
```
mkrefs apidocs docs/mkrefs.yml
mkrefs biblio docs/mkrefs.yml
mkrefs glossary docs/mkrefs.yml
```


## Caveats

While the [`MkDocs`](https://www.mkdocs.org/) utility is astoundingly useful,
its documentation (and coding style) leave much room for improvement.
The [documentation for developing plugins](https://www.mkdocs.org/user-guide/plugins/#developing-plugins)
is not even close to what happens when its code executes.

Consequently, the **MkRefs** project is an attempt to reverse-engineer
the code from many other MkDocs plugins, while documenting its observed
event sequence, required parameters, limitations and workarounds, etc.

Two issues persist, where you will see warnings even though the **MkRefs**
code is handling configuration as recommended:

```
WARNING -  Config value: 'mkrefs_config'. Warning: Unrecognised configuration name: mkrefs_config 
```

and

```
INFO    -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
  - biblio.md
  - glossary.md
  - ref.md
```

For now, you can simply ignore both of these warnings.
Meanwhile, we'll work on eliminating them.


## Feature roadmap

Let us know if you need features to parse and generate
[BibTeX](http://www.bibtex.org/).


## License and Copyright

Source code for **MkRefs** plus its logo, documentation, and examples
have an [MIT license](https://spdx.org/licenses/MIT.html) which is
succinct and simplifies use in commercial applications.

All materials herein are Copyright &copy; 2021 Derwen, Inc.


## Acknowledgements

Many thanks to our open source [sponsors](https://github.com/sponsors/ceteri);
and to our contributors:

changelog.txt
# `MkRefs` changelog

## 0.3.0

2021-07-??

  * migrated to `mkdocs-material` as part of development
  * fixed bug in RDF for function parameters
  * using `imporlib` approach to find local source module
  * fixed bug when apidocs is not configured

## 0.2.0

2021-06-27

  * added glossary support
  * added apidocs support


## 0.1.0

2021-05-22

  * initial release, bibliography only


mkrefs/apidocs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# see license https://github.com/DerwenAI/mkrefs#license-and-copyright

"""
Implementation of apidoc-ish documentation which generates actual
Markdown that can be used with MkDocs, and fits with Diátaxis design
principles for effective documentation.

Because the others really don't.

In particular, this library...

  * is aware of type annotations (PEP 484, etc.)
  * provides non-bassackwards parameter descriptions (eyes on *you*, GOOG)
  * handles forward references (prior to Python 3.8)
  * links to source lines in a Git repo
  * fixes bugs in `typing` and `inspect`
  * does not require use of a separate apidocs plugin
  * uses `icecream` for debugging
  * exists b/c Sphinx sucks

You're welcome.
"""

import importlib
import inspect
import os
import re
import sys
import traceback
import typing

from icecream import ic  # type: ignore # pylint: disable=E0401
import kglab
import pathlib
import rdflib  # type: ignore  # pylint: disable=E0401

from .util import render_reference


class PackageDoc:
    """
There doesn't appear to be any other Markdown-friendly docstring support in Python.

See also:

  * [PEP 256](https://www.python.org/dev/peps/pep-0256/)
  * [`inspect`](https://docs.python.org/3/library/inspect.html)
    """
    PAT_PARAM = re.compile(r"(    \S+.*\:\n(?:\S.*\n)+)", re.MULTILINE)
    PAT_NAME = re.compile(r"^\s+(.*)\:\n(.*)")
    PAT_FWD_REF = re.compile(r"ForwardRef\('(.*)'\)")


    def __init__ (
        self,
        package_name: str,
        git_url: str,
        class_list: typing.List[str],
        ) -> None:
        """
Constructor, to configure a `PackageDoc` object.

    package_name:
name of the Python package

    git_url:
URL for the Git source repository

    class_list:
list of the classes to include in the apidocs
        """
        self.package_name = package_name
        self.git_url = git_url
        self.class_list = class_list

        # hunt for the package
        spec = importlib.util.spec_from_file_location(self.package_name, self.package_name + "/__init__.py")

        #self.package_obj = sys.modules[self.package_name]
        self.package_obj = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(self.package_obj)  # type: ignore
        sys.modules[spec.name] = self.package_obj

        # prepare a file path prefix (to remove later, per file)
        pkg_path = os.path.dirname(inspect.getfile(self.package_obj))
        self.file_prefix = "/".join(pkg_path.split("/")[0:-1])

        self.md: typing.List[str] = [
            "# Reference: `{}` package".format(self.package_name),
            ]

        self.meta: dict = {
            "package": self.package_name,
            "git_url": self.git_url,
            "class": {},
            "function": {},
            "type": {},
        }


    def show_all_elements (
        self
        ) -> None:
        """
Show all possible elements from `inspect` for the given package, for
debugging purposes.
        """
        for name, obj in inspect.getmembers(self.package_obj):
            for n, o in inspect.getmembers(obj):
                ic(name, n, o)
                ic(type(o))


    def get_todo_list (
        self
        ) -> typing.Dict[ str, typing.Any]:
        """
Walk the package tree to find class definitions to document.

    returns:
a dictionary of class objects which need apidocs generated
        """
        todo_list: typing.Dict[ str, typing.Any] = {
            class_name:  class_obj
            for class_name, class_obj in inspect.getmembers(self.package_obj, inspect.isclass)
            if class_name in self.class_list
            }

        return todo_list


    def build (
        self
        ) -> None:
        """
Build the apidocs documentation as markdown.
        """
        todo_list: typing.Dict[ str, typing.Any] = self.get_todo_list()

        # markdown for top-level package description
        docstring = self.get_docstring(self.package_obj)
        self.md.extend(docstring)

        self.meta["docstring"] = "\n".join(docstring).strip()

        # find and format the class definitions
        for class_name in self.class_list:
            class_obj = todo_list[class_name]
            self.format_class(class_name, class_obj)

        # format the function definitions and types
        self.format_functions()
        self.format_types()


    def get_docstring (  # pylint: disable=W0102
        self,
        obj: typing.Any,
        parse: bool = False,
        arg_dict: dict = {},
        ) -> typing.List[str]:
        """
Get the docstring for the given object.

    obj:
class definition for which its docstring will be inspected and parsed

    parse:
flag to parse docstring or use the raw text; defaults to `False`

    arg_dict:
optional dictionary of forward references, if parsed

    returns:
list of lines of markdown
        """
        local_md: typing.List[str] = []
        raw_docstring = obj.__doc__

        if raw_docstring:
            docstring = inspect.cleandoc(raw_docstring)

            if parse:
                local_md.append(self.parse_method_docstring(obj, docstring, arg_dict))
            else:
                local_md.append(docstring)

            local_md.append("\n")

        return local_md


    def parse_method_docstring (
        self,
        obj: typing.Any,
        docstring: str,
        arg_dict: dict,
        ) -> str:
        """
Parse the given method docstring.

    obj:
class definition currently being inspected

    docstring:
input docstring to be parsed

    arg_dict:
optional dictionary of forward references

    returns:
parsed/fixed docstring, as markdown
        """
        local_md: typing.List[str] = []

        for chunk in self.PAT_PARAM.split(docstring):
            m_param = self.PAT_PARAM.match(chunk)

            if m_param:
                param = m_param.group()
                m_name = self.PAT_NAME.match(param)

                if m_name:
                    name = m_name.group(1).strip()

                    if name not in arg_dict:
                        code = obj.__code__
                        line_num = code.co_firstlineno
                        module = code.co_filename
                        raise Exception(f"argument `{name}` described at line {line_num} in {module} is not in the parameter list")

                    anno = self.fix_fwd_refs(arg_dict[name])
                    descrip = m_name.group(2).strip()

                    if name == "returns":
                        local_md.append("\n  * *{}* : `{}`  \n{}".format(name, anno, descrip))
                    elif name == "yields":
                        local_md.append("\n  * *{}* :  \n{}".format(name, descrip))
                    else:
                        local_md.append("\n  * `{}` : `{}`  \n{}".format(name, anno, descrip))
            else:
                chunk = chunk.strip()

                if len(chunk) > 0:
                    local_md.append(chunk)

        return "\n".join(local_md)


    def fix_fwd_refs (
        self,
        anno: str,
        ) -> typing.Optional[str]:
        """
Substitute the quoted forward references for a given package class.

    anno:
raw annotated type for the forward reference

    returns:
fixed forward reference, as markdown; or `None` if no annotation is supplied
        """
        results: list = []

        if not anno:
            return None

        for term in anno.split(", "):
            for chunk in self.PAT_FWD_REF.split(term):
                if len(chunk) > 0:
                    results.append(chunk)

        return ", ".join(results)


    def document_method (
        self,
        path_list: list,
        name: str,
        obj: typing.Any,
        func_kind: str,
        func_meta: dict,
        ) -> typing.Tuple[int, typing.List[str]]:
        """
Generate apidocs markdown for the given class method.

    path_list:
elements of a class path, as a list

    name:
class method name

    obj:
class method object

    func_kind:
function kind

    func_meta:
function metadata

    returns:
line number, plus apidocs for the method as a list of markdown lines
        """
        local_md: typing.List[str] = ["---"]

        # format a header + anchor
        frag = ".".join(path_list + [ name ])
        anchor = "#### [`{}` {}](#{})".format(name, func_kind, frag)
        local_md.append(anchor)

        func_meta["ns_path"] = frag

        # link to source code in Git repo
        code = obj.__code__
        line_num = code.co_firstlineno
        file = code.co_filename.replace(self.file_prefix, "")

        src_url = "[*\[source\]*]({}{}#L{})\n".format(self.git_url, file, line_num)  # pylint: disable=W1401
        local_md.append(src_url)

        func_meta["file"] = file
        func_meta["line_num"] = line_num

        # format the callable signature
        sig = inspect.signature(obj)
        arg_list = self.get_arg_list(sig)
        arg_list_str = "{}".format(", ".join([ a[0] for a in arg_list ]))

        local_md.append("```python")
        local_md.append("{}({})".format(name, arg_list_str))
        local_md.append("```")

        func_meta["arg_list_str"] = arg_list_str

        # include the docstring, with return annotation
        arg_dict: dict = {
            name.split("=")[0]: anno
            for name, anno in arg_list
            }

        arg_dict["yields"] = None

        ret = sig.return_annotation

        if ret:
            arg_dict["returns"] = self.extract_type_annotation(ret)

        arg_docstring = self.get_docstring(obj, parse=True, arg_dict=arg_dict)
        local_md.extend(arg_docstring)
        local_md.append("")

        func_meta["arg_dict"] = arg_dict
        func_meta["arg_docstring"] = "\n".join(arg_docstring).strip()

        return line_num, local_md


    def get_arg_list (
        self,
        sig: inspect.Signature,
        ) -> list:
        """
Get the argument list for a given method.

    sig:
inspect signature for the method

    returns:
argument list of `(arg_name, type_annotation)` pairs
        """
        arg_list: list = []

        for param in sig.parameters.values():
            #ic(param.name, param.empty, param.default, param.annotation, param.kind)

            if param.name == "self":
                pass
            else:
                if param.kind == inspect.Parameter.VAR_POSITIONAL:
                    name = "*{}".format(param.name)
                elif param.kind == inspect.Parameter.VAR_KEYWORD:
                    name = "**{}".format(param.name)
                elif param.default == inspect.Parameter.empty:
                    name = param.name
                else:
                    if isinstance(param.default, str):
                        default_repr = repr(param.default).replace("'", '"')
                    else:
                        default_repr = param.default

                    name = "{}={}".format(param.name, default_repr)

                anno = self.extract_type_annotation(param.annotation)
                arg_list.append((name, anno))

        return arg_list


    @classmethod
    def extract_type_annotation (
        cls,
        sig: inspect.Signature,
        ):
        """
Extract the type annotation for a given method, correcting `typing`
formatting problems as needed.

    sig:
inspect signature for the method

    returns:
corrected type annotation
        """
        type_name = str(sig)
        type_class = sig.__class__.__module__

        try:
            if type_class != "typing":
                if type_name.startswith("<class"):
                    type_name = type_name.split("'")[1]

            if type_name == "~AnyStr":
                type_name = "typing.AnyStr"
            elif type_name.startswith("~"):
                type_name = type_name[1:]

        except Exception:  # pylint: disable=W0703
            ic(type_name)
            traceback.print_exc()

        return type_name


    def document_type (
        self,
        path_list: list,
        name: str,
        obj: typing.Any,
        ) -> typing.List[str]:
        """
Generate apidocs markdown for the given type definition.

    path_list:
elements of a class path, as a list

    name:
type name

    obj:
type object

    returns:
apidocs for the type, as a list of lines of markdown
        """
        local_md: typing.List[str] = []

        type_meta: typing.Dict[str, str] = {}
        self.meta["type"][name] = type_meta

        # format a header + anchor
        frag = ".".join(path_list + [ name ])
        anchor = "#### [`{}` {}](#{})".format(name, "type", frag)
        local_md.append(anchor)

        type_meta["ns_path"] = frag

        # show type definition
        local_md.append("```python")
        local_md.append("{} = {}".format(name, obj))
        local_md.append("```")
        local_md.append("")

        type_meta["obj"] = repr(obj)

        return local_md


    @classmethod
    def find_line_num (
        cls,
        src: typing.Tuple[typing.List[str], int],
        member_name: str,
        ) -> int:
        """
Corrects for the error in parsing source line numbers of class methods that have decorators:
<https://stackoverflow.com/questions/8339331/how-to-get-line-number-of-function-with-without-a-decorator-in-a-python-module>

    src:
list of source lines for the class being inspected

    member_name:
name of the class member to locate

    returns:
corrected line number of the method definition
        """
        correct_line_num = -1

        for line_num, line in enumerate(src[0]):
            tokens = line.strip().split(" ")

            if tokens[0] == "def" and tokens[1] == member_name:
                correct_line_num = line_num

        return correct_line_num


    def format_class (
        self,
        class_name: str,
        class_obj: typing.Any,
        ) -> None:
        """
Format apidocs as markdown for the given class.

    class_name:
name of the class to document

    class_obj:
class object
        """
        self.md.append("## [`{}` class](#{})".format(class_name, class_name))  # pylint: disable=W1308

        docstring = class_obj.__doc__
        src = inspect.getsourcelines(class_obj)

        class_meta = {
            "docstring": docstring,
            "method": {},
            }

        self.meta["class"][class_name] = class_meta

        if docstring:
            # add the raw docstring for a class
            self.md.append(docstring)

        obj_md_pos: typing.Dict[int, typing.List[str]] = {}

        for member_name, member_obj in inspect.getmembers(class_obj):
            path_list = [self.package_name, class_name]

            if member_name.startswith("__") or not member_name.startswith("_"):
                if member_name not in class_obj.__dict__:
                    # inherited method
                    continue

                if inspect.isfunction(member_obj):
                    func_kind = "method"
                elif inspect.ismethod(member_obj):
                    func_kind = "classmethod"
                else:
                    continue

                func_meta: typing.Dict[str, str] = {}
                class_meta["method"][member_name] = func_meta

                _, obj_md = self.document_method(path_list, member_name, member_obj, func_kind, func_meta)
                line_num = self.find_line_num(src, member_name)
                obj_md_pos[line_num] = obj_md

        for _, obj_md in sorted(obj_md_pos.items()):
            self.md.extend(obj_md)


    def format_functions (
        self
        ) -> None:
        """
Walk the package tree, and for each function definition format its
apidocs as markdown.
        """
        self.md.append("---")
        self.md.append("## [package functions](#{})".format(self.package_name))

        for func_name, func_obj in inspect.getmembers(self.package_obj, inspect.isfunction):
            if not func_name.startswith("_"):
                func_meta: typing.Dict[str, str] = {}
                self.meta["function"][func_name] = func_meta

                _, obj_md = self.document_method([self.package_name], func_name, func_obj, "function", func_meta)
                self.md.extend(obj_md)


    def format_types (
        self
        ) -> None:
        """
Walk the package tree, and for each type definition format its apidocs
as markdown.
        """
        self.md.append("---")
        self.md.append("## [package types](#{})".format(self.package_name))

        for name, obj in inspect.getmembers(self.package_obj):
            if obj.__class__.__module__ == "typing":
                if not str(obj).startswith("~"):
                    obj_md = self.document_type([self.package_name], name, obj)
                    self.md.extend(obj_md)


    @classmethod
    def entity_template (
        cls,
        kg: kglab.KnowledgeGraph,
        node: kglab.RDF_Node,
        kind: kglab.RDF_Node,
        name: str,
        descrip: str,
        parent: typing.Optional[kglab.RDF_Node],
        ) -> None:
        """
Represent the given entity in RDF.

    kg:
graph object

    node:
entity node being represented

    kind:
RDF type of the entity

    name:
name of the entity

    descrip:
Markdown description from docstring

    parent:
parent node of the entity
        """
        kg.add(node, kg.get_ns("rdf").type, kind)
        kg.add(node, kg.get_ns("rdfs").label, rdflib.Literal(name, lang=kg.language))
        kg.add(node, kg.get_ns("dct").description, rdflib.Literal(descrip, lang=kg.language))

        if parent:
            kg.add(node, kg.get_ns("dct").isPartOf, parent)


    @classmethod
    def function_template (
        cls,
        kg: kglab.KnowledgeGraph,
        node: kglab.RDF_Node,
        meta: dict,
        ) -> None:
        """
Represent additional metadata for a function in RDF.

    kg:
graph object

    node:
entity node being represented

    meta:
additional metadata
        """
        kg.add(node, kg.get_ns("derw").apidocs_ns_path, rdflib.Literal(meta["ns_path"], lang=kg.language))
        kg.add(node, kg.get_ns("derw").apidocs_args, rdflib.Literal(meta["arg_list_str"], lang=kg.language))
        kg.add(node, kg.get_ns("derw").apidocs_file, rdflib.Literal(meta["file"], lang=kg.language))
        kg.add(node, kg.get_ns("derw").apidocs_line, rdflib.Literal(meta["line_num"], datatype=rdflib.XSD.integer))

        if "yields" in meta["arg_dict"] and meta["arg_dict"]["yields"]:
            kg.add(node, kg.get_ns("derw").apidocs_yields, rdflib.Literal(meta["arg_dict"]["yields"], lang=kg.language))

            if "returns" in meta["arg_dict"] and meta["arg_dict"]["returns"]:
                kg.add(node, kg.get_ns("derw").apidocs_returns, rdflib.Literal(meta["arg_dict"]["returns"], lang=kg.language))

        for param_name, param_type in meta["arg_dict"].items():
            if param_name not in ["yields", "returns"]:
                param_node = rdflib.URIRef("derw:apidocs:param:" + meta["ns_path"] + "." + re.sub("[^0-9a-zA-Z_]", "", param_name))
                kg.add(param_node, kg.get_ns("rdf").type, kg.get_ns("derw").PythonParam)
                kg.add(param_node, kg.get_ns("rdfs").label, rdflib.Literal(param_name, lang=kg.language))
                kg.add(param_node, kg.get_ns("derw").apidocs_type, rdflib.Literal(param_type, lang=kg.language))
                kg.add(param_node, kg.get_ns("dct").isPartOf, node)


    def get_rdf (
        self
        ) -> kglab.KnowledgeGraph:
        """
Generate an RDF graph from the apidocs descriptions.

    returns:
generated knowledge graph
        """
        kg = kglab.KnowledgeGraph(
            namespaces={
                "dct": "http://purl.org/dc/terms/",
                "derw": "https://derwen.ai/ns/v1#",
                "owl": "http://www.w3.org/2002/07/owl#",
                "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                "skos": "http://www.w3.org/2004/02/skos/core#",
                "xsd":  "http://www.w3.org/2001/XMLSchema#",
            })

        package_node = rdflib.URIRef(f"derw:apidocs:package:{self.package_name}")
        kg.add(package_node, kg.get_ns("dct").identifier, rdflib.URIRef(self.git_url))

        self.entity_template(
            kg,
            package_node,
            kg.get_ns("derw").PythonPackage,
            self.package_name,
            self.meta["docstring"],
            None,
        )

        for class_name, class_obj in self.meta["class"].items():
            class_node = rdflib.URIRef(f"derw:apidocs:class:{self.package_name}.{class_name}")

            self.entity_template(
                kg,
                class_node,
                kg.get_ns("derw").PythonClass,
                class_name,
                class_obj["docstring"],
                package_node,
            )

            for method_name, method_obj in class_obj["method"].items():
                method_node = rdflib.URIRef(f"derw:apidocs:method:{self.package_name}.{class_name}.{method_name}")

                self.entity_template(
                    kg,
                    method_node,
                    kg.get_ns("derw").PythonMethod,
                    method_name,
                    method_obj["arg_docstring"],
                    class_node,
                )

                self.function_template(
                    kg,
                    method_node,
                    method_obj,
                )

        for function_name, function_obj in self.meta["function"].items():
            function_node = rdflib.URIRef(f"derw:apidocs:function:{self.package_name}.{function_name}")

            self.entity_template(
                kg,
                function_node,
                kg.get_ns("derw").PythonMethod,
                function_name,
                function_obj["arg_docstring"],
                package_node,
            )

            self.function_template(
                kg,
                function_node,
                function_obj,
            )

        return kg


def render_apidocs (  # pylint: disable=R0914
    local_config: dict,
    template_path: pathlib.Path,
    markdown_path: pathlib.Path,
    ) -> typing.Dict[str, list]:
    """
Render the Markdown for an apidocs reference page, based on the given
Jinja2 template.

    local_config:
local configuration, including user-configurable includes/excludes

    template_path:
file path for Jinja2 template for rendering an apidocs reference page in MkDocs

    markdown_path:
file path for the rendered Markdown file

    returns:
rendered Markdown
    """
    groups: typing.Dict[str, list] = {}

    package_name = local_config["apidocs"]["package"]
    git_url = local_config["apidocs"]["git"]

    includes = [
        name.strip()
        for name in local_config["apidocs"]["includes"].split(",")
    ]

    pkg_doc = PackageDoc(
        package_name,
        git_url,
        includes,
        )

    try:
        # hardcore debug only:
        #pkg_doc.show_all_elements()

        # build the apidocs markdown
        pkg_doc.build()

        # render the JSON into Markdown using the Jinja2 template
        groups = {
            "package": [ pkg_doc.meta ],
        }

        render_reference(
            template_path,
            markdown_path,
            groups,
        )
    except Exception as e:  # pylint: disable=W0703
        print(f"Error rendering apidocs: {e}")
        traceback.print_exc()

    return groups
